File: app/src/main/java/eu/kanade/mangafeed/data/source/online/english/Kissmanga.java
package eu.kanade.mangafeed.data.source.online.english;

import android.content.Context;
import android.net.Uri;

import com.squareup.okhttp.FormEncodingBuilder;
import com.squareup.okhttp.Headers;
import com.squareup.okhttp.Response;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Locale;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import eu.kanade.mangafeed.data.database.models.Chapter;
import eu.kanade.mangafeed.data.database.models.Manga;
import eu.kanade.mangafeed.data.source.SourceManager;
import eu.kanade.mangafeed.data.source.base.Source;
import eu.kanade.mangafeed.data.source.model.MangasPage;
import eu.kanade.mangafeed.data.source.model.Page;
import rx.Observable;

public class Kissmanga extends Source {

    public static final String NAME = "Kissmanga (EN)";
    public static final String HOST = "kissmanga.com";
    public static final String IP = "93.174.95.110";
    public static final String BASE_URL = "http://" + IP;
    public static final String POPULAR_MANGAS_URL = BASE_URL + "/MangaList/MostPopular?page=%s";
    public static final String SEARCH_URL = BASE_URL + "/AdvanceSearch";

    public Kissmanga(Context context) {
        super(context);
    }

    @Override
    protected Headers.Builder headersBuilder() {
        Headers.Builder builder = super.headersBuilder();
        builder.add("Host", HOST);
        return builder;
    }

    @Override
    public String getName() {
        return NAME;
    }

    @Override
    public int getId() {
        return SourceManager.KISSMANGA;
    }

    @Override
    public String getBaseUrl() {
        return BASE_URL;
    }

    @Override
    public boolean isLoginRequired() {
        return false;
    }

    @Override
    protected String getInitialPopularMangasUrl() {
        return String.format(POPULAR_MANGAS_URL, 1);
    }

    @Override
    protected String getInitialSearchUrl(String query) {
        return SEARCH_URL;
    }

    @Override
    protected List<Manga> parsePopularMangasFromHtml(Document parsedHtml) {
        List<Manga> mangaList = new ArrayList<>();

        Elements mangaHtmlBlocks = parsedHtml.select("table.listing tr:gt(1)");
        for (Element currentHtmlBlock : mangaHtmlBlocks) {
            Manga currentManga = constructPopularMangaFromHtmlBlock(currentHtmlBlock);
            mangaList.add(currentManga);
        }

        return mangaList;
    }

    private Manga constructPopularMangaFromHtmlBlock(Element htmlBlock) {
        Manga mangaFromHtmlBlock = new Manga();
        mangaFromHtmlBlock.source = getId();

        Element urlElement = htmlBlock.select("td a:eq(0)").first();

        if (urlElement != null) {
            mangaFromHtmlBlock.setUrl(urlElement.attr("href"));
            mangaFromHtmlBlock.title = urlElement.text();
        }

        return mangaFromHtmlBlock;
    }

    @Override
    protected String parseNextPopularMangasUrl(Document parsedHtml, MangasPage page) {
        Element next = parsedHtml.select("li > a:contains(› Next)").first();
        if (next == null)
            return null;

        return BASE_URL + next.attr("href");
    }

    public Observable<MangasPage> searchMangasFromNetwork(MangasPage page, String query) {
        if (page.page == 1)
            page.url = getInitialSearchUrl(query);

        FormEncodingBuilder form = new FormEncodingBuilder();
        form.add("authorArtist", "");
        form.add("mangaName", query);
        form.add("status", "");
        form.add("genres", "");

        return networkService
                .postData(page.url, form.build(), requestHeaders)
                .flatMap(networkService::mapResponseToString)
                .map(Jsoup::parse)
                .doOnNext(doc -> page.mangas = parseSearchFromHtml(doc))
                .doOnNext(doc -> page.nextPageUrl = parseNextSearchUrl(doc, page, query))
                .map(response -> page);
    }

    @Override
    protected List<Manga> parseSearchFromHtml(Document parsedHtml) {
        return parsePopularMangasFromHtml(parsedHtml);
    }

    @Override
    protected String parseNextSearchUrl(Document parsedHtml, MangasPage page, String query) {
        return null;
    }

    @Override
    protected Manga parseHtmlToManga(String mangaUrl, String unparsedHtml) {
        Document parsedDocument = Jsoup.parse(unparsedHtml);

        Element infoElement = parsedDocument.select("div.barContent").first();
        Element titleElement = infoElement.select("a.bigChar").first();
        Element authorElement = infoElement.select("p:has(span:contains(Author:)) > a").first();
        Elements genreElement = infoElement.select("p:has(span:contains(Genres:)) > *:gt(0)");
        Elements descriptionElement = infoElement.select("p:has(span:contains(Summary:)) ~ p");
        Element thumbnailUrlElement = parsedDocument.select(".rightBox:eq(0) img").first();

        Manga newManga = new Manga();
        newManga.url = mangaUrl;

        if (titleElement != null) {
            newManga.title = titleElement.text();
        }
        if (authorElement != null) {
            newManga.author = authorElement.text();
        }
        if (descriptionElement != null) {
            newManga.description = descriptionElement.text();
        }
        if (genreElement != null) {
            newManga.genre = genreElement.text();
        }
        if (thumbnailUrlElement != null) {
            newManga.thumbnail_url = Uri.parse(thumbnailUrlElement.attr("src"))
                    .buildUpon().authority(IP).toString();
        }
//        if (statusElement != null) {
//            boolean fieldCompleted = statusElement.text().contains("Completed");
//            newManga.status = fieldCompleted + "";
//        }

        newManga.initialized = true;

        return newManga;
    }

    @Override
    protected List<Chapter> parseHtmlToChapters(String unparsedHtml) {
        Document parsedDocument = Jsoup.parse(unparsedHtml);

        List<Chapter> chapterList = new ArrayList<>();

        Elements chapterElements = parsedDocument.select("table.listing tr:gt(1)");
        for (Element chapterElement : chapterElements) {
            Chapter currentChapter = constructChapterFromHtmlBlock(chapterElement);

            chapterList.add(currentChapter);
        }

        return chapterList;
    }

    private Chapter constructChapterFromHtmlBlock(Element chapterElement) {
        Chapter newChapter = Chapter.create();

        Element urlElement = chapterElement.select("a").first();
        Element dateElement = chapterElement.select("td:eq(1)").first();

        if (urlElement != null) {
            newChapter.setUrl(urlElement.attr("href"));
            newChapter.name = urlElement.text();
        }
        if (dateElement != null) {
            try {
                newChapter.date_upload = new SimpleDateFormat("MM/dd/yyyy", Locale.ENGLISH).parse(dateElement.text()).getTime();
            } catch (ParseException e) {
                // Do Nothing.
            }
        }

        newChapter.date_fetch = new Date().getTime();

        return newChapter;
    }

    public Observable<List<Page>> pullPageListFromNetwork(final String chapterUrl) {
        FormEncodingBuilder builder = new FormEncodingBuilder();
        return networkService
                .postData(getBaseUrl() + overrideChapterUrl(chapterUrl), builder.build(), requestHeaders)
                .flatMap(networkService::mapResponseToString)
                .flatMap(unparsedHtml -> {
                    List<String> pageUrls = parseHtmlToPageUrls(unparsedHtml);
                    return Observable.just(getFirstImageFromPageUrls(pageUrls, unparsedHtml));
                });
    }

    @Override
    protected List<String> parseHtmlToPageUrls(String unparsedHtml) {
        Document parsedDocument = Jsoup.parse(unparsedHtml);
        List<String> pageUrlList = new ArrayList<>();

        int numImages = parsedDocument.select("#divImage img").size();

        for (int i = 0; i < numImages; i++) {
            pageUrlList.add("");
        }
        return pageUrlList;
    }

    @Override
    protected List<Page> getFirstImageFromPageUrls(List<String> pageUrls, String unparsedHtml) {
        List<Page> pages = convertToPages(pageUrls);

        Pattern p = Pattern.compile("lstImages.push\\(\"(.+?)\"");
        Matcher m = p.matcher(unparsedHtml);
        List<String> imageUrls = new ArrayList<>();
        while (m.find()) {
            imageUrls.add(m.group(1));
        }

        for (int i = 0; i < pages.size(); i++) {
            pages.get(i).setImageUrl(imageUrls.get(i));
        }
        return pages;
    }

    @Override
    protected String parseHtmlToImageUrl(String unparsedHtml) {
        return null;
    }

    @Override
    public Observable<Response> getImageProgressResponse(final Page page) {
        return networkService.getProgressResponse(page.getImageUrl(), null, page);
    }

}

File: app/src/main/java/eu/kanade/tachiyomi/source/online/english/Kissmanga.kt
package eu.kanade.tachiyomi.source.online.english

import com.squareup.duktape.Duktape
import eu.kanade.tachiyomi.network.GET
import eu.kanade.tachiyomi.network.POST
import eu.kanade.tachiyomi.source.model.*
import eu.kanade.tachiyomi.source.online.ParsedHttpSource
import okhttp3.FormBody
import okhttp3.Headers
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import org.jsoup.nodes.Document
import org.jsoup.nodes.Element
import timber.log.Timber
import java.text.SimpleDateFormat
import java.util.regex.Pattern

class Kissmanga : ParsedHttpSource() {

    override val id: Long = 4

    override val name = "Kissmanga"

    override val baseUrl = "http://kissmanga.com"

    override val lang = "en"

    override val supportsLatest = true

    override val client: OkHttpClient = network.cloudflareClient

    override fun headersBuilder(): Headers.Builder {
        return Headers.Builder()
            .add("User-Agent", "Mozilla/5.0 (Windows NT 10.0; WOW64) Gecko/20100101 Firefox/60")
    }

    override fun popularMangaSelector() = "table.listing tr:gt(1)"

    override fun latestUpdatesSelector() = "table.listing tr:gt(1)"

    override fun popularMangaRequest(page: Int): Request {
        return GET("$baseUrl/MangaList/MostPopular?page=$page", headers)
    }

    override fun latestUpdatesRequest(page: Int): Request {
        return GET("http://kissmanga.com/MangaList/LatestUpdate?page=$page", headers)
    }

    override fun popularMangaFromElement(element: Element): SManga {
        val manga = SManga.create()
        element.select("td a:eq(0)").first().let {
            manga.setUrlWithoutDomain(it.attr("href"))
            val title = it.text()
            //check if cloudfire email obfuscation is affecting title name
            if (title.contains("[email protected]", true)) {
                try {
                    var str: String = it.html()
                    //get the  number
                    str = str.substringAfter("data-cfemail=\"")
                    str = str.substringBefore("\">[email")
                    val sb = StringBuilder()
                    //convert number to char
                    val r = Integer.valueOf(str.substring(0, 2), 16)!!
                    var i = 2
                    while (i < str.length) {
                        val c = (Integer.valueOf(str.substring(i, i + 2), 16) xor r).toChar()
                        sb.append(c)
                        i += 2
                    }
                    //replace the new word into the title
                    manga.title = title.replace("[email protected]", sb.toString(), true)
                } catch (e: Exception) {
                    //on error just default to obfuscated title
                    Timber.e("error parsing [email protected]", e)
                    manga.title = title
                }
            } else {
                manga.title = title
            }
        }
        return manga
    }

    override fun latestUpdatesFromElement(element: Element): SManga {
        return popularMangaFromElement(element)
    }

    override fun popularMangaNextPageSelector() = "li > a:contains(› Next)"

    override fun latestUpdatesNextPageSelector(): String = "ul.pager > li > a:contains(Next)"

    override fun searchMangaRequest(page: Int, query: String, filters: FilterList): Request {
        val form = FormBody.Builder().apply {
            add("mangaName", query)

            for (filter in if (filters.isEmpty()) getFilterList() else filters) {
                when (filter) {
                    is Author -> add("authorArtist", filter.state)
                    is Status -> add("status", arrayOf("", "Completed", "Ongoing")[filter.state])
                    is GenreList -> filter.state.forEach { genre -> add("genres", genre.state.toString()) }
                }
            }
        }
        return POST("$baseUrl/AdvanceSearch", headers, form.build())
    }

    override fun searchMangaSelector() = popularMangaSelector()

    override fun searchMangaFromElement(element: Element): SManga {
        return popularMangaFromElement(element)
    }

    override fun searchMangaNextPageSelector() = null

    override fun mangaDetailsParse(document: Document): SManga {
        val infoElement = document.select("div.barContent").first()

        val manga = SManga.create()
        manga.author = infoElement.select("p:has(span:contains(Author:)) > a").first()?.text()
        manga.genre = infoElement.select("p:has(span:contains(Genres:)) > *:gt(0)").text()
        manga.description = infoElement.select("p:has(span:contains(Summary:)) ~ p").text()
        manga.status = infoElement.select("p:has(span:contains(Status:))").first()?.text().orEmpty().let { parseStatus(it) }
        manga.thumbnail_url = document.select(".rightBox:eq(0) img").first()?.attr("src")
        return manga
    }

    fun parseStatus(status: String) = when {
        status.contains("Ongoing") -> SManga.ONGOING
        status.contains("Completed") -> SManga.COMPLETED
        else -> SManga.UNKNOWN
    }

    override fun chapterListSelector() = "table.listing tr:gt(1)"

    override fun chapterFromElement(element: Element): SChapter {
        val urlElement = element.select("a").first()

        val chapter = SChapter.create()
        chapter.setUrlWithoutDomain(urlElement.attr("href"))
        chapter.name = urlElement.text()
        chapter.date_upload = element.select("td:eq(1)").first()?.text()?.let {
            SimpleDateFormat("MM/dd/yyyy").parse(it).time
        } ?: 0
        return chapter
    }

    override fun pageListRequest(chapter: SChapter) = POST(baseUrl + chapter.url, headers)

    override fun pageListParse(response: Response): List<Page> {
        val body = response.body()!!.string()

        val pages = mutableListOf<Page>()

        // Kissmanga now encrypts the urls, so we need to execute these two scripts in JS.
        val ca = client.newCall(GET("$baseUrl/Scripts/ca.js", headers)).execute().body()!!.string()
        val lo = client.newCall(GET("$baseUrl/Scripts/lo.js", headers)).execute().body()!!.string()

        Duktape.create().use {
            it.evaluate(ca)
            it.evaluate(lo)

            // There are two functions in an inline script needed to decrypt the urls. We find and
            // execute them.
            var p = Pattern.compile("(var.*CryptoJS.*)")
            var m = p.matcher(body)
            while (m.find()) {
                it.evaluate(m.group(1))
            }

            // Finally find all the urls and decrypt them in JS.
            p = Pattern.compile("""lstImages.push\((.*)\);""")
            m = p.matcher(body)

            var i = 0
            while (m.find()) {
                val url = it.evaluate(m.group(1)) as String
                pages.add(Page(i++, "", url))
            }
        }

        return pages
    }

    override fun pageListParse(document: Document): List<Page> {
        throw Exception("Not used")
    }

    override fun imageUrlRequest(page: Page) = GET(page.url)

    override fun imageUrlParse(document: Document) = ""

    private class Status : Filter.TriState("Completed")
    private class Author : Filter.Text("Author")
    private class Genre(name: String) : Filter.TriState(name)
    private class GenreList(genres: List<Genre>) : Filter.Group<Genre>("Genres", genres)

    override fun getFilterList() = FilterList(
            Author(),
            Status(),
            GenreList(getGenreList())
    )

    // $("select[name=\"genres\"]").map((i,el) => `Genre("${$(el).next().text().trim()}", ${i})`).get().join(',\n')
    // on http://kissmanga.com/AdvanceSearch
    private fun getGenreList() = listOf(
            Genre("4-Koma"),
            Genre("Action"),
            Genre("Adult"),
            Genre("Adventure"),
            Genre("Comedy"),
            Genre("Comic"),
            Genre("Cooking"),
            Genre("Doujinshi"),
            Genre("Drama"),
            Genre("Ecchi"),
            Genre("Fantasy"),
            Genre("Gender Bender"),
            Genre("Harem"),
            Genre("Historical"),
            Genre("Horror"),
            Genre("Josei"),
            Genre("Lolicon"),
            Genre("Manga"),
            Genre("Manhua"),
            Genre("Manhwa"),
            Genre("Martial Arts"),
            Genre("Mature"),
            Genre("Mecha"),
            Genre("Medical"),
            Genre("Music"),
            Genre("Mystery"),
            Genre("One shot"),
            Genre("Psychological"),
            Genre("Romance"),
            Genre("School Life"),
            Genre("Sci-fi"),
            Genre("Seinen"),
            Genre("Shotacon"),
            Genre("Shoujo"),
            Genre("Shoujo Ai"),
            Genre("Shounen"),
            Genre("Shounen Ai"),
            Genre("Slice of Life"),
            Genre("Smut"),
            Genre("Sports"),
            Genre("Supernatural"),
            Genre("Tragedy"),
            Genre("Webtoon"),
            Genre("Yaoi"),
            Genre("Yuri")
    )
}
